pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -181,6 +181,53 @@ def add_oa_cuts(
                         )
 
 
+def add_oa_cuts_for_grey_box(
+    target_model, jacobians_model, config, objective_sense, mip_iter, cb_opt=None
+):
+    sign_adjust = -1 if objective_sense == minimize else 1
+    if config.add_slack:
+        slack_var = target_model.MindtPy_utils.cuts.slack_vars.add()
+    for target_model_grey_box, jacobian_model_grey_box in zip(
+        target_model.MindtPy_utils.grey_box_list,
+        jacobians_model.MindtPy_utils.grey_box_list,
+    ):
+        jacobian_matrix = (
+            jacobian_model_grey_box.get_external_model()
+            .evaluate_jacobian_outputs()
+            .toarray()
+        )
+        for index, output in enumerate(target_model_grey_box.outputs.values()):
+            dual_value = jacobians_model.dual[jacobian_model_grey_box][
+                output.name.replace("outputs", "output_constraints")
+            ]
+            target_model.MindtPy_utils.cuts.oa_cuts.add(
+                expr=copysign(1, sign_adjust * dual_value)
+                * (
+                    sum(
+                        jacobian_matrix[index][var_index] * (var - value(var))
+                        for var_index, var in enumerate(
+                            target_model_grey_box.inputs.values()
+                        )
+                    )
+                )
+                - (output - value(output))
+                - (slack_var if config.add_slack else 0)
+                <= 0
+            )
+    # TODO: gurobi_persistent currently does not support greybox model.
+    if (
+        config.single_tree
+        and config.mip_solver == 'gurobi_persistent'
+        and mip_iter > 0
+        and cb_opt is not None
+    ):
+        cb_opt.cbLazy(
+            target_model.MindtPy_utils.cuts.oa_cuts[
+                len(target_model.MindtPy_utils.cuts.oa_cuts)
+            ]
+        )
+
+
```

Comment:
These cuts assume that the black box is: 1. in the objective, 2. is convex. For the 1st one, can we generalize it? For the second one, can we use an option to decide whether or not we want to generate the augmented penalty cuts? Our examples/tests satisfy both assumptions, but we can get some flexibility for not so much effort.

Reply:
In this implementation, we don't assume the grey box is in the objective function. It will also work if it is in the constraints. 
The constraints inside the grey box are nonlinear equality constraints. If the constraint is convex, we should use **Equality Relaxation**. If the constraint is nonconvex, we should use both **Equality Relaxation** and **Augmented Penalty**. We can add slack variables using `config.add_slack`, the same as algebraic constraints. 
Therefore, our current implementation works even without the two assumptions.